src/realm/object-store/sectioned_results.cpp

```diff
@@ -0,0 +1,262 @@
+    {
+        auto modified_sections = std::map<size_t, std::vector<size_t>>();
+        for (auto i : indicies) {
+            auto range = section_for_index(offsets, i);
+            auto it = std::find(range.indices.begin(), range.indices.end(), i);
+            if (it != range.indices.end()) {
+                auto index = std::distance(range.indices.begin(), it);
+                modified_sections[range.index].push_back(index);
+            }
+        }
+        return modified_sections;
+    }
+
+private:
+    SectionedResultsNotificatonCallback m_cb;
+    SectionedResults& m_sectioned_results;
+    std::vector<SectionRange> m_prev_offset_ranges;
+    util::Optional<size_t> m_section_filter;
+};
+
+Mixed ResultsSection::operator[](size_t idx) const
+{
+    m_parent->calculate_sections_if_required();
+    return m_parent->m_results.get_any(m_parent->m_offset_ranges[m_index].indices[idx]);
+}
+
+Mixed ResultsSection::key()
+{
+    return m_parent->m_offset_ranges[m_index].key;
+}
+
+template <typename Context>
+auto ResultsSection::get(Context& ctx, size_t row_ndx)
+{
+    m_parent->calculate_sections_if_required();
+    return m_parent->m_results.get(ctx, m_parent->m_offset_ranges[m_index].indices[row_ndx]);
+}
+
+size_t ResultsSection::size()
+{
+    m_parent->calculate_sections_if_required();
+    REALM_ASSERT(m_parent->m_offset_ranges.size() > m_index);
+    auto range = m_parent->m_offset_ranges[m_index];
+    return range.indices.size();
+}
+
+NotificationToken ResultsSection::add_notification_callback(SectionedResultsNotificatonCallback callback,
+                                                            KeyPathArray key_path_array) &
+{
+    return m_parent->add_notification_callback_for_section(m_index, std::move(callback), key_path_array);
+}
+
+SectionedResults::SectionedResults(Results results, ComparisonFunc comparison_func)
+    : m_results(results)
+    , m_callback(std::move(comparison_func))
+{
+    m_offset_ranges = calculate_sections(m_results, m_callback);
+}
+
+SectionedResults::SectionedResults(Results results, Results::SectionedResultsOperator op,
+                                   util::Optional<StringData> prop_name)
+    : m_results(results)
+    , m_callback(builtin_comparison(results, op, prop_name))
+{
+    m_offset_ranges = calculate_sections(m_results, m_callback);
+}
+
+void SectionedResults::calculate_sections_if_required(Results::EvaluateMode mode)
+{
+    util::CheckedUniqueLock lock(m_results.m_mutex);
+    // m_results.ensure_up_to_date may indicate that the underlying collection has not changed
+    // even though it may have changed. To solve this, query the collection directly with
+    // `m_results.get_collection()->has_changed()` to get the actual source of truth.
+    if ((m_results.get_collection() != nullptr) && !m_results.get_collection()->has_changed()) {
+        return;
+    }
+    else if ((m_results.get_collection() == nullptr) && m_results.ensure_up_to_date(mode)) {
```

Comment:
It might be simpler to store `m_results.get_table()->get_content_version()` and compare the current to the previous version to see if the underlying table has changed.